core/src/test/resources/org/apache/calcite/test/TopDownOptTest.xml

```diff
@@ -35,6 +35,136 @@ EnumerableLimit(fetch=[5])
   EnumerableSortedAggregate(group=[{3}], EXPR$1=[COUNT()])
     EnumerableSort(sort0=[$3], dir0=[DESC-nulls-last])
       EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+    </TestCase>
+    <TestCase name="testSortProject">
+        <Resource name="sql">
+            <![CDATA[select mgr from sales.emp order by mgr desc nulls last]]>
+        </Resource>
+        <Resource name="planBefore">
+            <![CDATA[
+LogicalSort(sort0=[$0], dir0=[DESC-nulls-last])
+  LogicalProject(MGR=[$3])
+    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+        <Resource name="planAfter">
+            <![CDATA[
+EnumerableProject(MGR=[$3])
+  EnumerableSort(sort0=[$3], dir0=[DESC-nulls-last])
+    EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+    </TestCase>
+    <TestCase name="testSortProjectWhenCastLeadingToNoTMonotonic">
+        <Resource name="sql">
+            <![CDATA[select deptno from sales.emp order by cast(deptno as varchar) desc]]>
+        </Resource>
+        <Resource name="planBefore">
+            <![CDATA[
+LogicalSort(sort0=[$1], dir0=[DESC])
+  LogicalProject(DEPTNO=[$7], EXPR$1=[CAST($7):VARCHAR NOT NULL])
+    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+        <Resource name="planAfter">
+            <![CDATA[
+EnumerableSort(sort0=[$1], dir0=[DESC])
+  EnumerableProject(DEPTNO=[$7], EXPR$1=[CAST($7):VARCHAR NOT NULL])
+    EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+    </TestCase>
+    <TestCase name="testSortProjectDerive">
+        <Resource name="sql">
+            <![CDATA[
+select * from
+(select ename, job, max_sal from
+(select ename, job, max(sal) as max_sal from sales.emp group by ename, job) t) r
+join sales.bonus s on r.job=s.job and r.ename=s.ename
+]]>
```

Comment:
Does this query want to test the trait derivation for Project?
If you remove the trait derivation code for Project, I guess the plan might be the same?

Reply:
Ah the plan is the same even if remove trait derivation implementation. Honestly I don't have an explanation (still learning top-down optimization).

The intention here is I explicitly add another project between the left sorted aggregation and join, to test if the sort collation is derived by Join thus Join can enforce right input has the same collation. (e.g. I add an extra `select ename, job, max_sal from` on top of the left aggregation).

Any comment or suggestion on how should a project trait derivation should be properly tested?

Reply:
The project is removable. We can change
`select ename, job, max_sal from` to `select ename, job, max_sal+1 from`

Reply:
We are almost done. Project's trait propagation is tricky, unlike filter.

Reply:
`max_sal+1` works like a charm.